proposals/js-promise-integration/Overview.md

```diff
@@ -2,299 +2,226 @@
+### Suspendable functions
 
-0. Let `func` be the function that is passed to the `WebAssembly.Function` constructor, 
-1. the function that is created using this variant of the `WebAssembly.Function` constructor will, when called with arguments `args`:
-1. Let `promise` be a new `Promise` constructed as though by the `Promise`(`fn`) constructor, where `fn` is a function of two arguments `accept` and `reject` that:
-   1. lets `suspender` be a new `Suspender` object and passes it as an additional argument to `args` to the `func` argument in the `WebAssembly.Function` constructor:
-      1. If `promising` is `"first"`, then `suspender` is the first argument, followed by `args`;
-      2. if `promising` is `"last"`, then `suspender` is the last argument, after `args`.
-   2. sets the state of `suspender` to **Active**[`caller`] (where `caller` is the current caller)
-   3. lets `result` be the result of calling `func(suspender,...args)` (or any trap or thrown exception)
-   4. asserts that `suspender`'s state is **Active**[`caller'`] for some `caller'` (should be guaranteed, though the caller might have changed)
-   5. changes `suspender`'s state to **Moribund**. This is also an opportunity to release any execution resources associated with the suspender. A **Moribund** suspender may not be used to suspend computations.
-   6. If `result` is not an exception or a trap, calls the `accept` function argument with the appropriate value.
-   6. If `result` is an exception, or if it is a trap, calls the `reject` function with the raised exception.
-6. Returns `promise` to `caller`
+We use the `Suspendable` constructor to signal to WebAssembly that a given import should cause a suspension of WebAssembly execution.
 
-Note that, if the inner function `func` suspends (by invoking a `Promise` returning import), then the `promise` will be returned to the `caller` before `func` returns. When `func` completes eventually, then `promise` will be resolved&mdash;and one of `accept` or `reject` will be invoked by the browser's microtask runner.
+#### `new WebAssembly.Suspending(`*`func`*`)`
 
-## Frequently Asked Questions
+The `WebAssembly.Suspending` constructor takes a JavaScript `Function` as an argument which is embedded in the `Suspending` object as the value of the hidden `wrappedFunction` property.
+
+>Note that *`func`* is expected to be a *JavaScript function*. This allows us to ignore certain so-called corner cases in the usage of JSPI: in particular there is no special handling of WebAssembly functions passed to `WebAssembly.Suspending`.
+
+1. If IsCallable(*func*) is `false`, throw a `TypeError` exception.
+1. Let *suspendingProto* be `WebAssembly.Suspendable.%prototype%`
+1. Let *susp* be the result of `OrdinaryObjectCreate`(*`suspendingProto`*)
+1. Perform `!CreateDataPropertyOrThrow`(*susp*,`"wrappedFunction"`,*func*)
```

Comment:
This creates a normal property on the object, not anything "hidden". I suspect what you want here is a JS internal slot, usually denoted `[[LikeThis]]`. There doesn't seem to be a way to talk about this in WebIDL, but there are web specs that use them, for example https://www.w3.org/TR/css-typed-om-1/#dom-stylepropertymapreadonly-declarations-slot

Reply:
Ah, I see. Unfortunately `[Unscopable]` only affects visibility inside a `with` statement (very niche; it was introduced to avoid some web compat hazards in the ES6 timeframe). The property is still attached to the object and visible during regular property lookups (`foo.propName`).

Reply:
Thanks. I did have some difficulty figuring out the correct annotation ...
To confirm, the declaration should look like:

Function [[wrappedFunction]];

Is that correct?

Reply:
There is no WebIDL annotation for this (see https://github.com/whatwg/webidl/issues/258 for some discussion and links to many other discussions). This can only be done in prose at the moment; the CSS Typed OM link above has some example usage, and Tab would probably be a good person to chat with for best practices.

Reply:
I have rewritten to use [[wrappedFunction]] + prose.

Reply:
Thanks, the use of the internal slot lgtm.

Reply:
Thanks, the use of the internal slot lgtm.

Reply:
'is of the form ..' is currently used in our spec. E.g., the 'name of the WebAssembly function' algorithm states:

If |funcinst| is of the form {type <var ignore>functype</var>, hostcode |hostfunc|},

and the ToJSValue algorithm states:

If |w| is of the form [=i64.const=] |i64|,

to quote two of many examples.